72_config_and_simplify_tls.md

```diff
@@ -0,0 +1,426 @@
+
+- [ ] Enhancement issue in release milestone, which links to pull request in
+  [keylime/enhancements]
+- [ ] Core members have approved the issue with the label `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+
+The `keylime.conf` is cleaned up and split into different sections.
+The new configuration format is versioned to allow for simpler upgrades in the
+future.
+
+The current TLS setup uses one CA for the verifier, one for the registrar and
+one for the agent mTLS authentication. This will be replaced by a client and
+server (key and certificate) option in each component (agent, verifier,
+registrar, tenant). The client pair is used to authenticate the component using
+mTLS and the server pair is used for TLS server (like the registrar, agent and
+verifier have). Also two new options are introduced, one for trustworthy
+incoming CAs for server side mTLS authentication and another one for trustworthy
+CAs from the client side.
+
+## Motivation
+
+Keylime uses the Python specific INI inspired configuration format and
+mixes different data types (comma separated list vs. Python syntax).
+Furthermore the configuration for different components are not separated
+and there is no nice way to upgrade users to newly introduced configuration
+options.
+
+The current way TLS is setup in Keylime is confusing to newcomers. Also the
+different configuration options are not consistent in their naming scheme.
+
+
+### Goals
+* Split the config into five new configuration files: global, agent, tenant,
+  registrar, verifier
+* Add versioning to the config
+* Provide a migration tool to port common Keylime configurations to the new
+  format
+* Simplify the TLS setup of Keylime
+* Cleanup keylime_ca to make it work with the new setup
+* Formally allow the Rust agent configuration to diverge from the Python agent
+
+### Non-Goals
+
+* Implement other mechanisms of authentication other than mTLS
+* Move registration to HTTPS (but can be easily done in a followup change)
+* Add a migration path for all possible old Keylime configurations
+
+## Proposal - Split the global configuration into the different components and Cleanup
+
+Distributions should ship the default configuration in `/usr/etc/keylime/`
+or if not supported in `/etc/keylime/`. If `/etc/keylime/` is populated the
+options in `/usr/etc/keylime` will be ignored. Manual overrides can be
+provided in `/etc/keylime.d/`. The Rust agent will adopt another
+configuration format and can diverge from the Python agent options (not all
+old options are useful for the rust agent).
+
+For the new configuration is split into the following files with the following
+options:
+
+* `global.conf`: Contains global configuration.
```

Comment:
Is this global configuration needed? What if we moved these to the individual configuration files?

I think the individual files should have their own version and probably it makes sense to allow different logging configuration for each service. Then the global configuration would keep only the `enable_tls` option which we are considering dropping.

Note that there is a good chance that the most used deployment is to have each service running in a separate machine or container.

Reply:
You are right, we can remove the global.conf. `enable_tls` is as you said currently broken, so we can remove it.

If we want to load only the configuration for the service in use then the change is a lot harder because some parts need to be refactored to not depend on each other anymore. (Main issues are the logging and tpm module).